File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CurrencyExponent.kt
package com.lextures.android.core.lms

object CurrencyExponent {
    private val zeroDecimalCurrencies = setOf("jpy")

    fun isZeroDecimal(currency: String): Boolean =
        zeroDecimalCurrencies.contains(currency.lowercase().trim())

    fun minorUnitFactor(currency: String): Int = if (isZeroDecimal(currency)) 1 else 100

    fun minorUnitsToMajorUnits(minor: Int, currency: String): Double =
        minor.toDouble() / minorUnitFactor(currency)

    fun majorUnitsToMinorUnits(major: Double, currency: String): Int =
        Math.round(major * minorUnitFactor(currency)).toInt()

    fun maxCatalogMinorUnits(currency: String): Int =
        if (isZeroDecimal(currency)) 99_999 else 9_999_999

    const val MAX_PRICE_MAJOR_ZERO_DECIMAL = 99_999.0
}

File: clients/ios/Lextures/Core/LMS/CurrencyExponent.swift
import Foundation

enum CurrencyExponent {
    private static let zeroDecimalCurrencies: Set<String> = ["jpy"]

    static func isZeroDecimal(_ currency: String) -> Bool {
        zeroDecimalCurrencies.contains(currency.lowercased().trimmingCharacters(in: .whitespacesAndNewlines))
    }

    static func minorUnitFactor(_ currency: String) -> Int {
        isZeroDecimal(currency) ? 1 : 100
    }

    static func minorUnitsToMajorUnits(_ minor: Int, currency: String) -> Double {
        Double(minor) / Double(minorUnitFactor(currency))
    }

    static func majorUnitsToMinorUnits(_ major: Double, currency: String) -> Int {
        Int((major * Double(minorUnitFactor(currency))).rounded())
    }

    static func maxCatalogMinorUnits(_ currency: String) -> Int {
        isZeroDecimal(currency) ? 99_999 : 9_999_999
    }

    static let maxPriceMajorZeroDecimal = 99_999.0
}